src/MenuCache.php

```diff
@@ -82,6 +82,8 @@ public function get_menu( $menu, $args ) {
 		if ( $this->should_cache_menu( $args ) && isset( $args->{$this->key_argument} ) ) {
 			$cached_menu = get_transient( $args->{$this->key_argument} );
 			if ( is_string( $cached_menu ) ) {
+				// We have found what we need, so we don't want any code to run after us (since any changes made here would be impossible to restore)
+				remove_all_filters( 'pre_wp_nav_menu' );
```

Comment:
While using the above action hook is safe and sound most of the times, you might do dangerous stuff if someone else is using the menu cache in an incorrect way. No idea if we should take this into account or not, but it could be done by checking if the current action actually is the expected one, or not...